elide-core/src/main/java/com/yahoo/elide/security/executors/ActivePermissionExecutor.java

```diff
@@ -158,46 +182,80 @@ public ActivePermissionExecutor(final com.yahoo.elide.core.RequestScope requestS
         if (commitExpression != null) {
             commitCheckQueue.add(new QueuedCheck(commitExpression, annotationClass));
         }
+        return ExpressionResult.DEFERRED;
     }
 
     /**
      * Check strictly user permissions on a specific field and entity.
      *
+     * @param <A> type parameter
      * @param resource Resource
      * @param annotationClass Annotation class
      * @param field Field
-     * @param <A> type parameter
      */
     @Override
-    public <A extends Annotation> void checkUserPermissions(PersistentResource<?> resource,
-                                                            Class<A> annotationClass,
-                                                            String field) {
+    public <A extends Annotation> ExpressionResult checkUserPermissions(PersistentResource<?> resource,
+                                                                        Class<A> annotationClass,
+                                                                        String field) {
         if (requestScope.getSecurityMode() == SecurityMode.SECURITY_INACTIVE) {
-            return; // Bypass
+            return ExpressionResult.PASS; // Bypass
+        }
+
+        // If the user check has already been evaluated before, return the result directly and save the building cost
+        ExpressionResult expressionResult
+                = userPermissionCheckCache.get(Triple.of(annotationClass, resource.getResourceClass(), field));
```

Comment:
Does the building cost save much?  If this cache was removed, do we see an appreciable slowdown?